server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java

```diff
@@ -325,6 +479,69 @@ public RecoverySettings(Settings settings, ClusterSettings clusterSettings) {
         );
     }
 
+    private void computeMaxBytesPerSec(Settings settings) {
+        // limit as computed before 8.1.0
+        final long defaultBytesPerSec = Math.max(INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING.get(settings).getBytes(), 0L);
+
+        // available network bandwidth
+        final long networkBandwidthBytesPerSec = Math.max(availableNetworkBandwidth.getBytes(), 0L);
+
+        // read bandwidth
+        final long readBytesPerSec;
+        if (availableDiskReadBandwidth.getBytes() > 0L && networkBandwidthBytesPerSec > 0L) {
+            double readFactor = NODE_BANDWIDTH_RECOVERY_FACTOR_READ_SETTING.get(settings);
+            readBytesPerSec = Math.round(Math.min(availableDiskReadBandwidth.getBytes(), networkBandwidthBytesPerSec) * readFactor);
+        } else {
+            readBytesPerSec = 0L;
+        }
+
+        // write bandwidth
+        final long writeBytesPerSec;
+        if (availableDiskWriteBandwidth.getBytes() > 0L && networkBandwidthBytesPerSec > 0L) {
+            double writeFactor = NODE_BANDWIDTH_RECOVERY_FACTOR_WRITE_SETTING.get(settings);
+            writeBytesPerSec = Math.round(Math.min(availableDiskWriteBandwidth.getBytes(), networkBandwidthBytesPerSec) * writeFactor);
+        } else {
+            writeBytesPerSec = 0L;
+        }
+
+        final long availableBytesPerSec = Math.min(readBytesPerSec, writeBytesPerSec);
+
+        long maxBytesPerSec;
+        if (availableBytesPerSec == 0L                                      // no node recovery bandwidths
+            || INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING.exists(settings)  // when set this setting overrides node recovery bandwidths
+            || DiscoveryNode.canContainData(settings) == false) {           // keep previous behavior for non data nodes
+            maxBytesPerSec = defaultBytesPerSec;
+        } else {
+            maxBytesPerSec = Math.max(defaultBytesPerSec, availableBytesPerSec);
+        }
+
+        final long maxAllowedBytesPerSec = Math.round(
+            availableBytesPerSec * NODE_BANDWIDTH_RECOVERY_OPERATOR_FACTOR_MAX_OVERCOMMIT_SETTING.get(settings)
```

Comment:
I wonder if this is what we want here? I would think this should be `min(availableDiskReadBandwidth, availableDiskWriteBandwidth, networkBandwidthBytesPerSec) * maxOverCommit`? At least in it's current form we practically only use this to limit the `indices.recovery.max_bytes_per_sec` setting so seems odd it should depend on the factors?

Reply:
I wasn't sure how we wanted to apply the overcommit and I do think it makes more sense like you suggest; I pushed 3f3890fd66c960146e6ff1dcca8278d512f176fc